src/numbuf/python/src/pynumbuf/memory.h

```diff
@@ -3,15 +3,63 @@
 
 #include <arrow/io/interfaces.h>
 
+/* C++ includes */
+#include <vector>
+#include <string>
+#include <thread>
+
+#define THREADPOOL_SIZE 8
+#define MEMCOPY_BLOCK_SIZE 64
+#define BYTES_IN_MB (1 << 20)
+
+using namespace std;
+
 namespace numbuf {
 
+void memcopy_aligned(uint8_t *dst, const uint8_t *src, uint64_t nbytes,
+                     uint64_t block_size, std::vector<std::thread>& threadpool) {
+  uint64_t num_threads = THREADPOOL_SIZE - 2;
+  uint64_t dst_address = reinterpret_cast<uint64_t>(dst);
+  uint64_t src_address = reinterpret_cast<uint64_t>(src);
+  uint64_t left_address = (src_address + block_size - 1) & ~(block_size - 1);
+  uint64_t right_address = (src_address + nbytes) & ~(block_size - 1);
+  uint64_t num_blocks = (right_address - left_address) / block_size;
+  // Update right address
+  right_address = right_address - (num_blocks % num_threads) * block_size;
+  // Now we divide these blocks between available threads. The remainder is
+  // handled on the main thread.
+  
+  uint64_t chunk_size = (right_address - right_address) / num_threads;
+  uint64_t prefix = left_address - src_address;
+  uint64_t suffix = src_address + nbytes - right_address;
+  // Now the data layout is | prefix | k * num_threads * block_size | suffix |.
+  // We have chunk_size = k * block_size, therefore the data layout is 
+  // | prefix | num_threads * chunk_size | suffix |.
+  // Each thread gets a "chunk" of k blocks.
+
+  // Start all the threads first. Then handle leftovers while threads run.
+  for (int i = 0; i < num_threads; i++) {
+    threadpool[i] =
+        std::thread(memcpy, dst + prefix + i * chunk_size, reinterpret_cast<uint8_t*>(left_address) + i * chunk_size, chunk_size);
+  }
+
+  memcpy(dst, src, prefix);
+  memcpy(dst + prefix + num_threads * chunk_size, reinterpret_cast<uint8_t*>(right_address), suffix);
+
+  for (auto &t: threadpool) {
+    if (t.joinable()) {
+      t.join();
+    }
+  }
+}
+
 class FixedBufferStream : public arrow::io::OutputStream,
                           public arrow::io::ReadableFileInterface {
  public:
   virtual ~FixedBufferStream() {}
 
   explicit FixedBufferStream(uint8_t* data, int64_t nbytes)
-      : data_(data), position_(0), size_(nbytes) {}
+      : data_(data), position_(0), size_(nbytes), threadpool_(THREADPOOL_SIZE) {}
```

Comment:
Hey Wes,
We're refactoring this into a standalone class. Should be easier to contribute to arrow as well when we're done with it.
Thanks,
Alexey